### PR TITLE
Read-Only mode in proxy

### PR DESCRIPTION
**Pull request by hysz**
## Description

This PR implements Read-Only mode for the Staking Contracts. There's a read-only proxy that emulates a static-delegate-call into the logic contract. Any state changes are reverted.

Enable this in the proxy by calling `StakingProxy.setReadOnlyMode(true)`.
Disable this in the proxy by calling `StakingProxy.setReadOnlyMode(false)`.

The nice thing about this implementation is that it does not depend on a read-only modifier.

To-dos:
1. Move the asm into a library
2. Revert when changing staking contract when in read-only mode.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/staking/contracts/src/ReadOnlyProxy.sol [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    /// @dev Executes a read-only call to the staking contract, via `revertDelegateCall`.
+    ///      By routing through `revertDelegateCall` any state changes are reverted.
+    // solhint-disable no-complex-fallback
+    function ()
+        external
+    {
+        address thisAddress = address(this);
+        bytes4 revertDelegateCallSelector = this.revertDelegateCall.selector;
+
+        assembly {
```
Comment (hysz) [inline contracts/staking/contracts/src/ReadOnlyProxy.sol]: (@todo - move this into a library DRY)

File: contracts/staking/contracts/src/ReadOnlyProxy.sol [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        }
+    }
+
+    /// @dev Executes a delegate call to the staking contract, if it is set.
+    ///      This function always reverts with the return data.
+    function revertDelegateCall()
+        external
+    {
+        address _readOnlyProxyCallee = readOnlyProxyCallee;
+        if (_readOnlyProxyCallee == address(0)) {
```
Comment (dorothy-zbornak) [inline contracts/staking/contracts/src/ReadOnlyProxy.sol]: Why even have this check? This function can't modify state.

File: contracts/staking/contracts/src/libs/LibProxy.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+*/
+
+pragma solidity ^0.5.9;
+
+import "@0x/contracts-utils/contracts/src/LibRichErrors.sol";
+import "./LibStakingRichErrors.sol";
+
+
+library LibProxy {
```
Comment (dorothy-zbornak) [inline contracts/staking/contracts/src/libs/LibProxy.sol]: This is such a cool library! I don't even care that the auditors will hate it. :trollface:

File: contracts/staking/contracts/src/libs/LibProxy.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    enum RevertRule {
+        REVERT_ON_ERROR,
+        ALWAYS_REVERT,
+        NEVER_REVERT
+    }
+
+    /// @dev Executes a read-only call to the staking contract, via `revertDelegateCall`.
+    ///      By routing through `revertDelegateCall` any state changes are reverted.
+    // solhint-disable no-complex-fallback
```
Comment (dorothy-zbornak) [inline contracts/staking/contracts/src/libs/LibProxy.sol]: Do we still need this linter toggle?

File: contracts/staking/contracts/src/libs/LibProxy.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    function proxyCall(
+        address destination,
+        RevertRule revertRule,
+        bytes4 customSelector,
+        bool ignoreSelector
+    )
+        internal
+    {
+        if (destination == address(0)) {
```
Comment (dorothy-zbornak) [inline contracts/staking/contracts/src/libs/LibProxy.sol]: I kind of feel like there's no point checking for this. A wrong address is just as bad as `0x0`.

File: contracts/staking/contracts/src/ReadOnlyProxy.sol (hunk 2 above)
Comment (hysz) [reply contracts/staking/contracts/src/ReadOnlyProxy.sol]: See the proxy lib, which now reverts if it is address zero. Reverting is important to distinguish between a legitimate NIL response and a misconfigured proxy.

File: contracts/staking/contracts/src/libs/LibStakingRichErrors.sol [hunk 6]
```diff
@@ -134,6 +134,10 @@ library LibStakingRichErrors {
     bytes4 internal constant INVALID_STAKE_STATUS_ERROR_SELECTOR =
         0xb7161acd;
 
+    // bytes4(keccak256("ProxyDestinationCannotBeNil()"))
+    bytes4 internal constant PROXY_DESTINATION_CANNOT_BE_NIL =
```
Comment (mzhu25) [inline contracts/staking/contracts/src/libs/LibStakingRichErrors.sol]: I think we're trying to use this optimization for rich errors that don't have any parameters:
https://github.com/0xProject/0x-monorepo/blob/88e56356c424cc834948891ecb0918ced7c56bb9/contracts/exchange-libs/contracts/src/LibMathRichErrors.sol#L8

File: contracts/staking/test/catastrophy_test.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import { BigNumber } from '@0x/utils';
+import { LogWithDecodedArgs } from 'ethereum-types';
+import * as _ from 'lodash';
+
+import { StakingProxyReadOnlyModeSetEventArgs } from '../src';
+
+import { StakingWrapper } from './utils/staking_wrapper';
+
+// tslint:disable:no-unnecessary-type-assertion
+blockchainTests.resets('Catastrophy Tests', () => {
```
Comment (mzhu25) [inline contracts/staking/test/catastrophy_test.ts]:
```suggestion
blockchainTests.resets('Catastrophe Tests', () => {
```
😆

Comment (hysz) [reply contracts/staking/test/catastrophy_test.ts]: Hahahahaha this word is my nemesis!

File: contracts/staking/contracts/src/libs/LibProxy.sol [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        ALWAYS_REVERT,
+        NEVER_REVERT
+    }
+
+    /// @dev Executes a read-only call to the staking contract, via `revertDelegateCall`.
+    ///      By routing through `revertDelegateCall` any state changes are reverted.
+    function proxyCall(
+        address destination,
+        RevertRule revertRule,
+        bytes4 customSelector,
```
Comment (abandeali1) [inline contracts/staking/contracts/src/libs/LibProxy.sol]: Is it correct to say that if you specify a non-zero `customSelector` then `ignoreSelector` will always be `true`?

Comment (hysz) [reply contracts/staking/contracts/src/libs/LibProxy.sol]: No—but is the distinction clear if they’re renamed to `customEgressSelector` and `ignoreIngressSelector`?

File: contracts/staking/contracts/src/libs/LibProxy.sol (hunk 5 above)
Comment (hysz) [reply contracts/staking/contracts/src/libs/LibProxy.sol]: That's true... I guess this is more of an initialization sanity check than a correctness check.

Comment (hysz) [reply contracts/staking/contracts/src/libs/LibProxy.sol]: We just had a cheeky chat about this in the coder cave. I think we're going to leave these sanity checks in for now (they're used throughout staking) but remove them if we do end up using constants.

File: contracts/staking/contracts/src/libs/LibProxy.sol (hunk 8 above)
Comment (hysz) [reply contracts/staking/contracts/src/libs/LibProxy.sol]: For reference:
```
                                                             | customEgressSelector | ignoreIngressSelector
StakingProxy fallback                        | no                                    | no
ReadOnlyProxy fallback                    | yes                                   | no
ReadOnlyProxy revertDelegateCall  | no                                    | yes
```

File: contracts/staking/test/catastrophe_test.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
+import { ERC20ProxyContract, ERC20Wrapper } from '@0x/contracts-asset-proxy';
+import { DummyERC20TokenContract } from '@0x/contracts-erc20';
+import { blockchainTests, describe, expect, provider, web3Wrapper } from '@0x/contracts-test-utils';
```
Comment (mzhu25) [inline contracts/staking/test/catastrophe_test.ts]: with blockchainTests we no longer need to import `provider` and `web3Wrapper`. we have these in the `BlockchainTestsEnvironment`

**Pull request merged**